include/zephyr/dt-bindings/pinctrl/ch32v003-pinctrl.h

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2024 Michael Hope
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __CH32V003_PINCTRL_H__
+#define __CH32V003_PINCTRL_H__
+
+/* Pin number 0-15 */
+#define CH32V003_PINCTRL_PIN_SHIFT   0
+/* If set, pin is an output of some type */
+#define CH32V003_PINCTRL_OUTPUT_BIT  3
+/* Port number with 0 = PA */
+#define CH32V003_PINCTRL_PORT_SHIFT  4
+/* Bit number of the first AFIO PCFR1. 0x1F means unset. */
+#define CH32V003_PINCTRL_AFIO0_SHIFT 6
+/* Bit number of the second AFIO PCFR1. 0x1F means unset. */
+#define CH32V003_PINCTRL_AFIO1_SHIFT 11
+#define CH32V003_PINCTRL_AFIO_UNSET  0x1F
+
+#define SPI1_NSS_PC1_0	   0xffe9
```

Comment:
I think it would be cleaner to use a define to generate these values, something like this

```
#define OUTPUT 1
#define INPUT 0

#define CH32V003_PINCTRL(pin, output, port, pcfr_bit0, pcfr_bit1)                               \
    ((pin & CH32V003_PINCTRL_PIN_SHIFT_MASK) << CH32V003_PINCTRL_PIN_SHIFT) |                   \
    ((output & CH32V003_PINCTRL_OUTPUT_BIT_MASK) << CH32V003_PINCTRL_OUTPUT_BIT) |              \
    ((((#port)[0]-'A') & CH32V003_PINCTRL_PORT_SHIFT_MASK) << CH32V003_PINCTRL_PORT_SHIFT) |    \
    ((pcfr_bit0 & CH32V003_PINCTRL_AFIO0_SHIFT_MASK) << CH32V003_PINCTRL_AFIO0_SHIFT) |         \
    ((pcfr_bit1 & CH32V003_PINCTRL_AFIO1_SHIFT_MASK) << CH32V003_PINCTRL_AFIO1_SHIFT)

#define SPI1_NSS_PC1_0	   \
    CH32V003_PINCTRL(1, OUTPUT, C, CH32V003_PINCTRL_AFIO_UNSET, CH32V003_PINCTRL_AFIO_UNSET)
```

Note that the formatting probably doesn't adhere to zephyr standards.

Reply:
I'm also looking into doing this for the CH32V208. I've even taken a slightly different approach where I define the PCFR value and the PCFR bit0 location and bit1 location as I feel it's a bit more readable. It looks something like this:

```
#define CH32V208_PINMUX(port, num, output, pcfr_val, pcfr_bit0, pcfr_bit1)                         \
	((PORT_TO_IDX(port) & PIN_PORT_MASK) << PIN_PORT_SHIFT) |                                  \
		((num & PIN_NUM_MASK) << PIN_NUM_SHIFT) |                                          \
		((output & PIN_OUTPUT_MASK) << PIN_OUTPUT_SHIFT) |                                 \
		((pcfr_val & PCFR_VALUE_MASK) << PCFR_VALUE_SHIFT) |                               \
		((pcfr_bit0 & PCFR_BIT0_STRUCT_OFFSET_MASK) << PCFR_BIT0_STRUCT_OFFSET_SHIFT) |    \
		((pcfr_bit1 & PCFR_BIT1_STRUCT_OFFSET_MASK) << PCFR_BIT1_STRUCT_OFFSET_SHIFT)

#define SPI3_DEFAULT_MAP 0
#define SPI3_REMAP       1

#define SPI3_NSS_PA15_0  CH32V208_PINMUX(A, 15, OUTPUT, SPI3_DEFAULT_MAP, 28, PCFR_BIT_UNUSED)
#define SPI3_SCK_PB3_0   CH32V208_PINMUX(B, 3, OUTPUT, SPI3_DEFAULT_MAP, 28, PCFR_BIT_UNUSED)
#define SPI3_MISO_PB4_0  CH32V208_PINMUX(B, 4, INPUT, SPI3_DEFAULT_MAP, 28, PCFR_BIT_UNUSED)
#define SPI3_MOSI_PB5_0  CH32V208_PINMUX(B, 5, OUTPUT, SPI3_DEFAULT_MAP, 28, PCFR_BIT_UNUSED)
#define SPI3_NSS_PA4_1   CH32V208_PINMUX(A, 4, OUTPUT, SPI3_REMAP, 28, PCFR_BIT_UNUSED)
#define SPI3_SCK_PC10_1  CH32V208_PINMUX(C, 10, OUTPUT, SPI3_REMAP, 28, PCFR_BIT_UNUSED)
#define SPI3_MISO_PC11_1 CH32V208_PINMUX(C, 11, INPUT, SPI3_REMAP, 28, PCFR_BIT_UNUSED)
#define SPI3_MOSI_PC12_1 CH32V208_PINMUX(C, 12, OUTPUT, SPI3_REMAP 28, PCFR_BIT_UNUSED)
```